File: Array/2.Medium/03.Majority_element.cpp
/*
QUESTION:-
Given an array nums of size n, return the majority element.
The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.

Example 1:

Input: nums = [3,2,3]
Output: 3

Example 2:

Input: nums = [2,2,1,1,1,2,2]
Output: 2
*/


#include <bits/stdc++.h>
using namespace std;
/*
APROACH:-
-> Initialize two variables: candidate and count. Set candidate to the first element of the array, and count to 1.
-> Iterate through the array starting from the second element:
    If the current element is equal to the candidate, increment the count by 1.
    If the current element is different from the candidate, decrement the count by 1.
    If the count becomes 0, update the candidate to the current element and set the count to 1 again.
-> After the iteration, the candidate variable will hold the majority element.
Return the candidate as the result.
*/

// CODE:-
int majorityElement(vector<int> &nums)
{
    int candidate = nums[0];
    int vote = 1;
    for (int i = 1; i < nums.size(); i++)
    {
        if (vote <= 0)
            candidate = nums[i];
        if (nums[i] == candidate)
            vote++;
        else
            vote--;
    }
    return candidate;
}


int main() {
    // Test case 1
    cout << "Test 1:" << endl;
    vector<int> nums1 = {3,2,3};
    int result1 = majorityElement(nums1);
    cout << "Output: " << result1 << endl;
    cout << "Expected: 3" << endl;
    cout << endl;

    // Test case 2
    cout << "Test 2:" << endl;
    vector<int> nums2 = {2,2,1,1,1,2,2};
    int result2 = majorityElement(nums2);
    cout << "Output: " << result2 << endl;
    cout << "Expected: 2" << endl;
    cout << endl;

    return 0;
}

// TIME COMPLEXITY = O(N)

File: 01.Arrays/2.Medium/03.Majority_element.py
# QUESTION:-
# Given an array nums of size n, return the majority element.
# The majority element is the element that appears more than floor(n/2) times.
# Example 1: Input: nums = [3,2,3], Output: 3
# Example 2: Input: nums = [2,2,1,1,1,2,2], Output: 2

# APPROACH:-
# Use Boyer-Moore Voting Algorithm
# Initialize candidate = first element, vote = 1
# For each element:
#   If vote == 0, update candidate to current element
#   If element == candidate, vote++, else vote--
# Return candidate

# CODE:-
def majorityElement(nums):
    """Find the majority element using Boyer-Moore Voting Algorithm"""
    candidate = nums[0]
    vote = 1
    
    for i in range(1, len(nums)):
        if vote == 0:
            candidate = nums[i]
            vote = 1
        elif nums[i] == candidate:
            vote += 1
        else:
            vote -= 1
    
    return candidate


if __name__ == '__main__':
    # Test case 1
    print("Test 1:")
    nums1 = [3, 2, 3]
    result1 = majorityElement(nums1)
    print(f"Output: {result1}")
    print("Expected: 3")
    print()
    
    # Test case 2
    print("Test 2:")
    nums2 = [2, 2, 1, 1, 1, 2, 2]
    result2 = majorityElement(nums2)
    print(f"Output: {result2}")
    print("Expected: 2")

# TIME COMPLEXITY = O(N)
# SPACE COMPLEXITY = O(1)
